### PR TITLE
Semver 2.0 support for VS 2015

### PR DESCRIPTION
**Pull request by nkolev92**
Resolves https://github.com/NuGet/Home/issues/5560

There are the 3 PRs that were ported:
https://github.com/NuGet/NuGet.Client/pull/1094

https://github.com/NuGet/NuGet.Client/pull/1128

https://github.com/NuGet/NuGet.Client/pull/1325

Fixed the inconsistencies and added some tests for that.

Now NuGet-3.6.0 will hit the "RegistrationBaseUrl/3.6.0" which was added in https://github.com/NuGet/NuGet.Services.Index/pull/70

**Note**

This PR is ready, but it's just a matter of how we work with the branches/versions.
We'll wait for @rrelyea before pulling the trigger on either this PR and the version changes currently in the base (release-3.5.1-rtm)

### CODE REVIEW

**Comment by joelverhagen** [comment]
Why did we port https://github.com/NuGet/NuGet.Client/pull/1094? It's not necessary for this to work, right?

**Comment by nkolev92** [comment]
@joelverhagen 
We could've done without it, but I there should be no harm in doing this. 
If this ships as 3.6.0 as we're proposing then it'd be nice to have the ability to make use of the versioning feature.

**Review by joelverhagen — APPROVED** [review]
`RegistrationBaseUrl/3.6.0` is not in PROD yet. Please let me know when these details are finalized so we can get it on `PROD`.

**Comment by nkolev92** [comment]
@joelverhagen perfect, thanks. Will ping you when ready.

**Comment by nkolev92** [comment]
Retargetted the branch. 
The release-3.5.1-rtm and release-3.6.0-rtm branches are identical (just different commit history)

**Review by emgarten — APPROVED** [review]

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/ServiceIndexResourceV3.cs [hunk 1]
```diff
@@ -14,129 +16,210 @@ namespace NuGet.Protocol
     /// </summary>
     public class ServiceIndexResourceV3 : INuGetResource
     {
-        private readonly IDictionary<string, List<Uri>> _index;
+        private readonly IDictionary<string, List<ServiceIndexEntry>> _index;
         private readonly DateTime _requestTime;
-        private List<Uri> _empty;
+        private static readonly IReadOnlyList<ServiceIndexEntry> _emptyEntries = new List<ServiceIndexEntry>();
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/ServiceIndexResourceV3.cs:21]: suggestion: Some comments on each of these would be nice to have. Else it is harder to understand.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/ServiceIndexResourceV3.cs [hunk 2]
```diff
@@ -14,129 +16,210 @@ namespace NuGet.Protocol
...
+        {
+            if (clientVersion == null)
+            {
+                throw new ArgumentNullException(nameof(clientVersion));
+            }
+
+            foreach (var type in orderedTypes)
+            {
+                List<ServiceIndexEntry> entries;
+                if (_index.TryGetValue(type, out entries))
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/ServiceIndexResourceV3.cs:73]: not: `List<ServiceIndexEntry> entries` do not need a separate declaration. they can be part of the function call. I can give you an example if needed.

File: test/NuGet.Core.Tests/NuGet.Protocol.VisualStudio.Tests/JsonData.cs [hunk 3]
```diff
@@ -1,19 +1,44 @@
...
     {
-        #region PSAutoCompleteV2Example
-        public const string PSAutoCompleteV2Example = @"[""elm.TypeScript.DefinitelyTyped"",""elmah"",""Elmah.AzureTableStorage"",""Elmah.BlogEngine.Net"",""Elmah.Contrib.EntityFramework"",""Elmah.Contrib.Mvc"",""Elmah.Contrib.WebApi"",""elmah.corelibrary"",""elmah.corelibrary.ews"",""Elmah.ElasticSearch"",""Elmah.Everywhere"",""elmah.ews"",""Elmah.FallbackErrorLog"",""elmah.filtering.sample"",""elmah.io"",""elmah.io.client"",""elmah.io.core"",""elmah.io.log4net"",""elmah.io.umbraco"",""elmah.mongodb"",""elmah.msaccess"",""Elmah.MVC"",""Elmah.MVC.ews"",""Elmah.MVC.XMLLight"",""elmah.mysql"",""elmah.oracle"",""elmah.postgresql"",""Elmah.RavenDB"",""Elmah.RavenDB.3"",""Elmah.RavenDB-4.5""]";
-        #endregion PSAutoCompleteV2Example
-
-
-        #region PSAutoCompleteV3Example
-        public const string PsAutoCompleteV3Example = @"{
+        public const string AutoCompleteV2Example = @"[
+    ""elm.TypeScript.DefinitelyTyped"",
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Protocol.VisualStudio.Tests/JsonData.cs:9]: 👍

**Review by mishra14 — APPROVED** [review]
Small comments. But Good Job otherwise! 🚢

**Pull request merged**